android/src/main/java/com/reactlibrary/RNCarnivalModule.java

```diff
@@ -40,524 +38,518 @@
+                    break;
+                case String:
+                    object.put(key, readableMap.getString(key));
+                    break;
+                case Map:
+                    object.put(key, convertMapToJson(readableMap.getMap(key)));
+                    break;
+                case Array:
+                    object.put(key, convertArrayToJson(readableMap.getArray(key)));
+                    break;
+            }
         }
-      });
-    } catch (URISyntaxException e) {
-      promise.reject(ERROR_CODE_TRACKING, e.getMessage());
+        return object;
     }
-  }
-
-  @ReactMethod
-  public void trackPageview(String url, ReadableArray tags, final Promise promise) {
-    try {
-      List<String> convertedTags = null;
-      if(tags != null) {
-        convertedTags = new ArrayList<>();
-        for (int i = 0; i < tags.size(); i++) {
-          convertedTags.add(tags.getString(i));
+
+    private static JSONArray convertArrayToJson(ReadableArray readableArray) throws JSONException {
+        JSONArray array = new JSONArray();
+        for (int i = 0; i < readableArray.size(); i++) {
+            switch (readableArray.getType(i)) {
+                case Null:
+                    break;
+                case Boolean:
+                    array.put(readableArray.getBoolean(i));
+                    break;
+                case Number:
+                    array.put(readableArray.getDouble(i));
+                    break;
+                case String:
+                    array.put(readableArray.getString(i));
+                    break;
+                case Map:
+                    array.put(convertMapToJson(readableArray.getMap(i)));
+                    break;
+                case Array:
+                    array.put(convertArrayToJson(readableArray.getArray(i)));
+                    break;
+            }
         }
-      }
+        return array;
+    }
+
+    @ReactMethod
+    public void setUserId(String userId) {
+        Carnival.setUserId(userId, null);
+    }
 
-      Carnival.trackPageview(new URI(url), convertedTags, new Carnival.TrackHandler() {
-        @Override
-        public void onSuccess() {
-          promise.resolve(true);
+    @ReactMethod
+    public void setUserEmail(String userEmail) {
+        Carnival.setUserEmail(userEmail, null);
+    }
+
+    @ReactMethod
+    public void getUnreadCount(Promise promise) {
+        int unreadCount = Carnival.getUnreadMessageCount();
+        promise.resolve(unreadCount);
+    }
+
+    @ReactMethod
+    public void removeMessage(ReadableMap messageMap) {
+        Message message = getMessage(messageMap);
+        Carnival.deleteMessage(message, null);
+    }
+
+    @ReactMethod
+    public void registerMessageImpression(int typeCode, ReadableMap messageMap) {
+        Message message = getMessage(messageMap);
+        CarnivalImpressionType type = null;
+
+        if (typeCode == 0) type = CarnivalImpressionType.IMPRESSION_TYPE_IN_APP_VIEW;
```

Comment:
What about changing to a switch case? https://docs.oracle.com/javase/tutorial/java/nutsandbolts/switch.html